stdlib/public/Concurrency/Task.cpp

```diff
@@ -1056,12 +1056,22 @@ swift_task_addCancellationHandlerImpl(
   auto *record = new (allocation)
       CancellationNotificationStatusRecord(unsigned_handler, context);
 
-  if (swift_task_addStatusRecord(record))
-    return record;
-
-  // else, the task was already cancelled, so while the record was added,
-  // we must run it immediately here since no other task will trigger it.
-  record->run();
+  bool fireHandlerNow = false;
+
+  swift_task_addStatusRecordWithChecks(record,
+                                       [&](ActiveTaskStatus parentStatus) {
+                                         if (parentStatus.isCancelled()) {
+                                           fireHandlerNow = true;
+                                           /* We don't fire the cancellation
+                                            * handler here since this function
+                                            * needs to be idempotent */
+                                         }
+                                         return true;
+                                       });
```

Comment:
As a general rule, we format around lambdas as if they were a sub-statement of the call, so this would be more like:

```
bool fireHandlerNow = false;
swift_task_addStatusRecordWithChecks(record, [&](ActiveTaskStatus parentStatus) {
  fireHandlerNow = parentStatus.isCancelled();
  return true; // Add the status record regardless
});
```